django/core/files/uploadhandler.py

```diff
@@ -138,15 +145,25 @@ def new_file(self, *args, **kwargs):
         """
         super().new_file(*args, **kwargs)
         self.file = TemporaryUploadedFile(self.file_name, self.content_type, 0, self.charset, self.content_type_extra)
+        self.uploaded = False
 
     def receive_data_chunk(self, raw_data, start):
         self.file.write(raw_data)
 
     def file_complete(self, file_size):
         self.file.seek(0)
         self.file.size = file_size
+        self.uploaded = True
```

Comment:
IMO this state should be controlled by the `MultiPartParser` not by a handler, e.g.
```diff
diff --git a/django/http/multipartparser.py b/django/http/multipartparser.py
index b3472f7be2..7fd7f0b49c 100644
--- a/django/http/multipartparser.py
+++ b/django/http/multipartparser.py
@@ -150,6 +150,7 @@ class MultiPartParser:
         num_post_keys = 0
         # To limit the amount of data read from the request.
         read_size = None
+        uploaded_file = True
 
         try:
             for item_type, meta_data, field_stream in Parser(stream, self._boundary):
@@ -159,6 +160,7 @@ class MultiPartParser:
                     # we hit the next boundary/part of the multipart content.
                     self.handle_file_complete(old_field_name, counters)
                     old_field_name = None
+                    uploaded_file = True
 
                 try:
                     disposition = meta_data['content-disposition'][1]
@@ -206,6 +208,7 @@ class MultiPartParser:
 
                     self._post.appendlist(field_name, force_str(data, encoding, errors='replace'))
                 elif item_type == FILE:
+                    uploaded_file = False
                     # This is a file, use the handler...
                     file_name = disposition.get('filename')
                     if file_name:
@@ -275,10 +278,15 @@ class MultiPartParser:
                     # If this is neither a FIELD or a FILE, just exhaust the stream.
                     exhaust(stream)
         except StopUpload as e:
+            for handler in handlers:
+                handler.upload_stopped()
             self._close_files()
             if not e.connection_reset:
                 exhaust(self._input_data)
         else:
+            if not uploaded_file:
+                for handler in handlers:
+                    handler.upload_stopped()
             # Make sure that the request data is all fed
             exhaust(self._input_data)
```

Reply:
Agreed. But I can't figure out, how we could test this functionality? Handling this state in the `UploadHandler`, just required making a custom handler and simulate the state of a situation that an interrupted upload would be in. I can't seem to understand how we could simulate a cancelled request condition in order to test the logic.